x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -530,7 +520,90 @@ public IndicesAccessControl authorize(
                 )
             );
         }
-        return new IndicesAccessControl(overallGranted, unmodifiableMap(indexPermissions));
+        return unmodifiableMap(indexPermissions);
+    }
+
+    /**
+     * Returns {@code true} if action is granted for all {@code requestedResources}.
+     * If action is not granted for at least one resource, this method will return {@code false}.
+     */
+    private boolean isActionGranted(final String action, final Map<String, IndexResource> requestedResources) {
+
+        if (requestedResources.isEmpty()) {
+            return true;
+        }
+
+        final boolean isMappingUpdateAction = isMappingUpdateAction(action);
+
+        for (IndexResource resource : requestedResources.values()) {
+            // true if ANY group covers the given index AND the given action
+            boolean granted = false;
+            // true if ANY group, which contains certain ingest privileges, covers the given index AND the action is a mapping update for
+            // an index or an alias (but not for a data stream)
+            boolean bwcGrantMappingUpdate = false;
+            final List<Runnable> bwcDeprecationLogActions = new ArrayList<>();
+
+            for (Group group : groups) {
+                // the group covers the given index OR the given index is a backing index and the group covers the parent data stream
+                if (resource.checkIndex(group)) {
+                    boolean actionCheck = group.checkAction(action);
+                    granted = granted || actionCheck;
```

Comment:
Since we're in the performance realm with this PR, we could shortcircuit the check here:
```
if (actionCheck) {
  granted = true;
  break;
}
```

Reply:
That's a good suggestion. Thanks for suggesting it. 🙏 

Analogously, do we want to stop at first `bwcMappingActionCheck==true` or is there any benefit from logging all deprecated privileges?

Reply:
Suggestion applied in commit: https://github.com/elastic/elasticsearch/pull/88662/commits/5dc47b770abd99431a93573faf7f494f29b39868